lib/SILOptimizer/Utils/CFGOptUtils.cpp

```diff
@@ -700,7 +701,73 @@ void swift::completeJointPostDominanceSet(
 
   // Now that we are done, add all remaining must visit blocks to our result
   // list. These are the remaining parts of our joint post-dominance closure.
-  llvm::copy(mustVisitSuccessorBlocks, std::back_inserter(result));
+  llvm::copy(mustVisitSuccessorBlocks, std::back_inserter(completion));
+
+  // Also, compute the postdominance set, which is the union of the user blocks
+  // that are not visited in the above fix point iteration and the
+  // mustVisitSuccessorBlocks.
+  llvm::copy_if(userBlocks, std::back_inserter(postDominators),
+                [&](SILBasicBlock *block) -> bool {
+                  return !visitedBlocks.count(block);
+                });
+  postDominators.append(completion.begin(), completion.end());
+}
+
+void swift::computePostDominatorsOfUsers(
+    SmallPtrSetImpl<SILInstruction *> &users,
+    SmallPtrSetImpl<SILInstruction *> &defs,
+    SmallVectorImpl<SILInstruction *> &postDominators) {
+  // Step 1. Collect the basic blocks of the users and defs.
+  SmallPtrSet<SILBasicBlock *, 4> userBlocks;
+  for (auto *inst : users) {
+    userBlocks.insert(inst->getParent());
+  }
+  SmallPtrSet<SILBasicBlock *, 4> defBlocks;
+  for (auto *inst : defs) {
+    defBlocks.insert(inst->getParent());
+  }
+
+  // Step 2. Compute post-dominators of all user blocks.
+  SmallVector<SILBasicBlock *, 8> completions;
+  SmallVector<SILBasicBlock *, 8> postDominatingBlocks;
+  completeJointPostDominanceSet(userBlocks, defBlocks, completions,
+                                postDominatingBlocks);
+
+  // Step 3. Compute all post-dominating instructions by looking into the
+  // post-dominating blocks.
+  SmallPtrSet<SILBasicBlock *, 8> completionSet(completions.begin(),
+                                                completions.end());
+  for (SILBasicBlock *bb : postDominatingBlocks) {
+    if (completionSet.count(bb)) {
+      // Here, the basic block bb was added to complete the postdominator set
+      // and does not have any users of defs. So, the first instruction of the
+      // block should be a part of the postdominator.
+      postDominators.push_back(&(bb->front()));
+      continue;
+    }
+
+    // Here, bb is a block containing one or more users and the post-dominating
+    // instruction is within this bb. It is the instruction following the last
+    // user in bb.
+    SILInstruction *instrAfterLastUser = nullptr;
+    for (auto currI = bb->rbegin(); currI != bb->rend(); currI++) {
+      SILInstruction *inst = &(*currI);
+      if (users.count(inst)) {
+        break;
+      }
+      instrAfterLastUser = inst;
+    }
+    if (instrAfterLastUser) {
+      postDominators.push_back(instrAfterLastUser);
+      continue;
+    }
+    // Here, the terminator of the basic block is a user. Therefore, use the
+    // first instruction of the successor blocks as post-dominating
+    // instructions.
+    for (SILBasicBlock *succ : bb->getSuccessors()) {
```

Comment:
Please assert that `succ` is not in `postDominators`.

Reply:
Sure.